keps/sig-storage/2644-honor-pv-reclaim-policy/README.md

```diff
@@ -0,0 +1,321 @@
+	if !ctrl.hasPVDeletionProtectionFinalizer(volume, pvutil.PVDeletionProtectionFinalizer) {
+		if newVolume.GetDeletionTimestamp() != nil {
+			klog.V(3).Infof("Volume %q is already being deleted", volume.Name)
+			return "", nil
+		}
+	}
+	//.....
+	//.....
+	//.....
+	klog.V(4).Infof("deleteVolumeOperation [%s]: success", volume.Name)
+        return pluginName, err
+    }
+// Remaining code below skipped for brevity...
+// Please refer pkg/controller/volume/persistentvolume/pv_controller.go in kubernetes/kubernetes for the full code.
+```
+
+On the driver side, the library adds the finalizer only to newly provisioned PVs. The proposal is to extend the library to (optionally) add the finalizer to all PVs (that are handled by the external-provisioner), to have all PVs protected once the feature is enabled.
+
+When the `shouldDelete` checks succeed, a delete volume request is initiated on the driver. This ensures that the volume is deleted on the backend.
+
+Once the volume is deleted from the backend, the finalizer can be removed. This allows the pv to be removed from the api server.
+
+Note: This feature should work with CSI Migration disabled or enabled.
+
+### Test Plan
+
+#### E2E tests
+An e2e test to exercise deletion of PV prior to deletion of PVC for a `Bound` PV-PVC pair.
+
+### Graduation Criteria
+
+#### Alpha
+* Feedback
+* Fix of the issue behind a feature flag.
+* Unit tests and e2e tests outlined in design proposal implemented.
+* Tests should be done with both CSI Migration disabled and enabled.
+
+#### Alpha -> Beta
+* Allow the Alpha fix to soak for one release.
+
+#### Beta -> GA
+* Deployed in production and have gone through at least one K8s upgrade.
+
+### Upgrade / Downgrade Strategy
+* Upgrade from old Kubernetes(1.22) to new Kubernetes(1.23) with `HonorPVReclaimPolicy` flag enabled
+
+In the above scenario, the upgrade will cause change in default behavior as described in the current KEP. Additionally,
+if there are PVs that have a valid associated PVC and deletion timestamp set, then a finalizer is added to the PV.
+
+* Downgrade from new Kubernetes(1.23) to old Kubernetes(1.22).
+
+In this case, there may be PVs with the deletion finalizer that the older Kubernetes does not remove. Such PVs will be in the API server forever unless if the user manually removes them.
+
+### Version Skew Strategy
+The fix is part of `kube-controller-manager` of `external-provisioner`.
+
+1. `kube-controller-manager` is upgraded, but `external-provisioner` is not:
+
+In this case the changed behavior is observed only for in-tree plugins. The drivers would still have the issue. Users
+will observe the new annotation on the PV, but it does not have any effect since the `external-provisioner` is not
+updated
+
+2. `external-provisioner` is upgraded but `kube-controller-manager`
+
+In this case the annotations are not added, the upgraded `external-provisioner` will have no effect, the issue would
+still exist.
```

Comment:
In this case the finalizer will be added and removed by the external-provisioner. PVs backed by the in-tree volume plugin are not protected.

In addition, PVs migrated to CSI have the finalizer. When the CSI migration is disabled, in-tree volume plugin / controller-manager does not remove the finalizer. The finalizer must be manually removed by the cluster admin.

Reply:
Fixed.